src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,421 @@
+        public DefaultSkipTokenHandler()
+        {
+        }
+
+        /// <summary>
+        /// Constructor for Unit testing purposes - Sets the Property Delimiter
+        /// </summary>
+        public DefaultSkipTokenHandler(char delimiter)
+            : this()
+        {
+            propertyDelimiter = delimiter;
+        }
+
+        /// <summary>
+        /// Returns the URI for NextPageLink
+        /// </summary>
+        /// <param name="baseUri">BaseUri for nextlink. It should be request URI for top level resource and navigationlink for nested resource.</param>
+        /// <param name="instance">Instance based on which SkipToken value will be generated.</param>
+        /// <param name="pageSize">Maximum number of records in the set of partial results for a resource.</param>
+        /// <param name="context">Serializer context</param>
+        /// <returns></returns>
+        public override Uri GenerateNextPageLink(Uri baseUri, int pageSize, Object instance, ODataSerializerContext context)
+        {
+            if (context == null)
+            {
+                return null;
+            }
+
+            if (pageSize <= 0)
+            {
+                return null;
+            }
+
+            Func<object, string> skipTokenGenerator = null;
+            IList<OrderByNode> orderByNodes = null;
+            ExpandedReferenceSelectItem expandedItem = context.ExpandedSelectItem;
+            IEdmModel model = context.Model;
+
+            DefaultQuerySettings settings = context.QueryOptions.Context.DefaultQuerySettings;
+            if (settings.EnableSkipToken)
+            {
+                if (expandedItem != null)
+                {
+                    // Handle Delta resource; currently not value based.
+                    if (TypedDelta.IsDeltaOfT(context.ExpandedResource.GetType()))
+                    {
+                        return GetNextPageHelper.GetNextPageLink(baseUri, pageSize);
+                    }
+
+                    if (expandedItem.OrderByOption != null)
+                    {
+                        orderByNodes = OrderByNode.CreateCollection(expandedItem.OrderByOption);
+                    }
+
+                    skipTokenGenerator = (obj) =>
+                    {
+                        return GenerateSkipTokenValue(obj, model, orderByNodes);
+                    };
+
+                    return GetNextPageHelper.GetNextPageLink(baseUri, pageSize, instance, skipTokenGenerator);
+                }
+
+                if (context.QueryOptions != null && context.QueryOptions.OrderBy != null)
+                {
+                    orderByNodes = context.QueryOptions.OrderBy.OrderByNodes;
+                }
+
+                skipTokenGenerator = (obj) =>
+                {
+                    return GenerateSkipTokenValue(obj, model, orderByNodes);
+                };
+            }
+            return context.InternalRequest.GetNextPageLink(pageSize, instance, skipTokenGenerator);
```

Comment:
>GetNextPageLink(pageSize, instance, skipTokenGenerator) [](start = 43, length = 55)

This seems to be the only place we call this new method.  Instead, could we call GetNextPageHelper directly, and remove the new overload from IWebAPIRequestMessage/WebAPIRequestMessage?

Reply:
We need to have GetNextPageLink as a method on the WebAPI request so that folks not using the EnableQuery attribute can create a paged result and set the nextlink through the request. If we takeaway this method then customers will not be able to access our implementation of next page link without using enable query attribute. The reason why this looks weird is because we only allowed customers to set the top-level nextlink and the nested nextlink would be taken care of by our serialization. 

---
In reply to: [287531231](https://github.com/OData/WebApi/pull/1724#discussion_r287531231) [](ancestors = 287531231)